llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -7579,16 +7579,23 @@ static bool isGuaranteedNotToBeUndefOrPoison(
     if (isa<UndefValue>(C))
       return !includesUndef(Kind);
 
-    if (isa<ConstantInt>(C) || isa<GlobalVariable>(C) || isa<ConstantFP>(V) ||
+    if (isa<ConstantInt>(C) || isa<GlobalVariable>(C) || isa<ConstantFP>(C) ||
         isa<ConstantPointerNull>(C) || isa<Function>(C))
       return true;
 
-    if (C->getType()->isVectorTy() && !isa<ConstantExpr>(C)) {
-      if (includesUndef(Kind) && C->containsUndefElement())
-        return false;
-      if (includesPoison(Kind) && C->containsPoisonElement())
-        return false;
-      return !C->containsConstantExpression();
+    if (C->getType()->isVectorTy()) {
+      if (isa<ConstantExpr>(C)) {
+        // Scalable vectors can use a ConstantExpr to build a splat.
+        if (Constant *SplatC = C->getSplatValue())
+          if (isa<ConstantInt>(SplatC))
+            return true;
```

Comment:
Any reason not to support fp/ptr splats?
```suggestion
          return isGuaranteedNotToBeUndefOrPoison(SplatC, ...);
```

Reply:
> Any reason not to support fp/ptr splats?

Just lack of tests. Should we increase the depth on the recursion or keep the current depth.

Reply:
The latter one is ok.

Reply:
I added ConstantFP and an Attributor test was changed so that has some coverage.

Will a splat of ConstantPointerNull be folded to ConstantAggregateZero at creation?